resolver/resolver.go

```diff
@@ -3507,33 +3494,85 @@ func (r *Resolver) createCELEnv(msg *Message) (*cel.Env, error) {
 	return env, nil
 }
 
-func (r *Resolver) getEnvMessage(msg *Message) *Message {
+func (r *Resolver) buildEnvMessage(msg *Message, svcMsgSet map[*Service]map[*Message]struct{}) (*Message, error) {
 	if msg.File == nil {
-		return nil
+		return nil, nil
 	}
 	if msg.File.Package == nil {
-		return nil
+		return nil, nil
 	}
-	var envSvcs []*Service
-	for _, file := range msg.File.Package.Files {
-		for _, svc := range file.Services {
-			if svc.Rule == nil {
-				continue
-			}
-			if svc.Rule.Env == nil {
-				continue
+
+	// Examine all the services where the message is used and collect all the essential information
+	// to calculate the intersection set of each service's environment variables.
+	svcEnvVarMap := map[*Service]map[string]*EnvVar{}
+	envVarNameSet := map[string]struct{}{}
+	for svc, msgSet := range svcMsgSet {
+		if _, exist := msgSet[msg]; exist {
+			envVarMap := make(map[string]*EnvVar)
+			if svc.Rule != nil && svc.Rule.Env != nil {
+				for _, envVar := range svc.Rule.Env.Vars {
+					envVarMap[envVar.Name] = envVar
+					envVarNameSet[envVar.Name] = struct{}{}
+				}
 			}
-			envSvcs = append(envSvcs, svc)
+			svcEnvVarMap[svc] = envVarMap
 		}
 	}
 
-	// If there are multiple services with env, we do not have a method to select one from them.
-	// Therefore, we proceed only when there is a single candidate service.
-	// Since validation for cases with multiple services is done elsewhere, we don't need to create an error here.
-	if len(envSvcs) != 1 {
-		return nil
+	// Calculate the intersection set of each service's environment variables
+	var envVars []*EnvVar
+	var errs []error
+	for envVarName := range envVarNameSet {
+		omit := false
+		type svcEnvVar struct {
+			svc    *Service
+			envVar *EnvVar
+		}
+		var svcEnvVars []*svcEnvVar
+		for svc, envVarMap := range svcEnvVarMap {
+			envVar, exist := envVarMap[envVarName]
+			if !exist {
+				// The given envVar is not defined in the service, so just omit it from the intersection set
+				omit = true
+				break
+			}
+			svcEnvVars = append(svcEnvVars, &svcEnvVar{svc: svc, envVar: envVar})
+		}
+		if omit {
+			continue
+		}
+
+		// Check if all the envVar has the same type
+		typ := svcEnvVars[0].envVar.Type
+		var mismatch bool
+		for _, sev := range svcEnvVars[1:] {
+			if !isExactlySameType(typ, sev.envVar.Type) {
+				mismatch = true
+				break
+			}
+		}
+		if mismatch {
+			var attrs []string
+			for _, sev := range svcEnvVars {
+				attrs = append(attrs, fmt.Sprintf(
+					"service: %s, type: %v, repeated: %v",
+					sev.svc.Name,
+					sev.envVar.Type.Kind.ToString(),
+					sev.envVar.Type.Repeated),
+				)
+			}
+			// Fix the output order mainly for testing
+			sort.Strings(attrs)
+			errs = append(errs, fmt.Errorf("environment variable %q has different types across services: %s", envVarName, strings.Join(attrs, ", ")))
+		}
+
+		envVars = append(envVars, svcEnvVars[0].envVar)
+	}
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
```

Comment:
It might become difficult to read if there are multiple errors, so it seems better to add each error to the context using addError.